SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestRepNotify/SpatialTestRepNotify.cpp

```diff
@@ -256,11 +265,65 @@ void ASpatialTestRepNotify::PrepareTest()
 			FinishStep();
 		},
 		nullptr, 5.0f);
+
+	/**
+	 * In this section of the test, we check the order rep notifies are called in.
```

Comment:
This section can currently fail since we don't have component set updates in yet, right?

Reply:
Though it should be very rare

Reply:
Yes. If they aren't on the same spatial component then they won't be together.

Reply:
I don't know how rare it will actually be though and it's best not to have test that you know flake. Am I right in thinking this test works almost as well if the components are all regular data components on the same object though?

Reply:
yeah - I figured the runtime would very very rarely be overloaded during tests (I got the impression that's when this would fail). I'll put a comment in. Hmmm, I was just thinking we could change the category of the test, but repnotifies are pretty core functionality. Is it worth moving this into its own test?

Reply:
Also commented that before Github loaded your comments Sami. Most of the test does yes - but the parts I've added are testing that data updates are applied to both the subobject and the actor before repnotifies are called (which we can put into some component set when we have component set updates)

Reply:
this passed 100/100 times, seems like we're not hitting the load needed for it to fail. seems like probably worth popping it in so we have it for when we have component set updates

Reply:
It's not overload that causes this it's just a random timing issue (although being overloaded can increase the probability).
Essentially we send two updates to the runtime (or more) and it sends them down to workers that are interested. 
The updates are sent separately and fanned out separately and so will be received separately on the other worker.

If the worker happens to tick between these two then it will cause the race.

However n PIE workers all run one after the other, so the chance of this happening is less.

Reply:
If we merge it in I doubt it would be a disaster. The options that we have are to merge it in and disable it if it starts to flake or add it in when we add in component set updates. However we don't know exactly when this will be.

Reply:
We could also get too keen and have some flake detection. If neither of them see the other state set then we know we hit the flake.

Reply:
ah makes sense, I guess I assumed some artifact of implementation meant it only happened when overloaded.

I guess we have a few options
1. move this test into a separate branch so it doesn't go into master until we have component set updates (and hopefully we remember about it)
2. include the test in the PR and if we do see it fail CI take it out (this depends on the % chance we think it will fail, if it's 0.001% the extra test coverage might be worth it
3. move this test into a nightly category, miron is of the opinion we can have some tests that may possibly fail in here, though I hear you disagree
4. This gets me thinking, my understanding is we do have guarantees about having all updates on actor creation, so could make a separate test actor and subobject at same time, update properties, then we don't have this chance to fail?

Reply:
Oh, I see your new comments, not sure about the flake detection, could be an option if we repeated it multiple times in the same test I guess (then report an actual failure if it fails too many times)

Tbh happy to take any approach, 4. sounds like it would be the most robust for now, but the most work

Reply:
Hmm actor creation is an interesting idea. And the answer is that yes we do have some guarantees around that. For the strongest guarantees if they are standard data (not owner only or server only) and on the same SpatialOS entity then they should be observed together in all cases.

Reply:
I'm generally of the opinion that tests should be correct with the only possibly flakes being that they time out (and those should be rare). Although in this case we expect we are going to support it soon.

I do like the idea of making it an entity creation thing if we want to separate out the idea that we are testing stuff is atomic from testing that rep notifies happen together.

Really the best way to test that rep notifies happen together is to have explicitly constructed fake updates to apply and see how they react. Then the whole thing is really a unit test. However I'm not sure we have what we need to do that quickly.

Reply:
Aight, will give the entity creation thing a go